# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' @title Find Isoforms
#' @description Find isoforms for given miRNAs and reads.
#' @param mirnas A DataFrame containing miRNA information. The DataFrame should
#'   include the following columns:
#'   \itemize{
#'     \item mature_id: Unique identifier of the mature miRNA.
#'     \item mature_seq: Sequence of the mature miRNA.
#'     \item seed_seq: Seed sequence of the mature miRNA.
#'     \item template_seq: Template sequence including flanking regions.
#'     \item flank_5p_seq: 5' flanking sequence of the mature miRNA.
#'     \item flank_3p_seq: 3' flanking sequence of the mature miRNA.
#'   }
#' @param reads A DataFrame containing read information. The DataFrame should
#'   include the following columns:
#'   \itemize{
#'     \item read_seq: Sequence of the read.
#'     \item tpm: TPM of the read.
#'   }
#' @param max_indel_5p An integer specifying the maximum allowed indels at the
#'   5' end. Must be non-negative.
#' @param max_indel_3p An integer specifying the maximum allowed indels at the
#'   3' end. Must be non-negative.
#' @param max_snp_5p An integer specifying the maximum allowed mismatches at
#'   the 5' end. Must be non-negative.
#' @param max_snp_3p An integer specifying the maximum allowed mismatches at
#'   the 3' end. Must be non-negative.
#' @param max_snp_seed An integer specifying the maximum allowed mismatches in
#'   the seed region. Must be non-negative.
#' @param max_snp An integer specifying the maximum allowed mismatches. Must be
#'   non-negative.
#' @return A DataFrame containing the detected isoforms with the following columns:
#'   \itemize{
#'     \item mature_id: Reference miRNA ID.
#'     \item mature_seq: Reference miRNA sequence.
#'     \item seed_seq: Seed sequence of the mature miRNA.
#'     \item template_seq: Template sequence including flanking regions.
#'     \item read_seq: Read sequence matched to the reference.
#'     \item read_num: Abundance of the read.
#'     \item tpm: TPM of the read.
#'     \item indel_5p: Indels at the 5' end.
#'     \item indel_3p: Indels at the 3' end.
#'     \item snp_5p: Mismatches at the 5' end.
#'     \item snp_3p: Mismatches at the 3' end.
#'     \item snp_seed: Mismatches in the seed.
#'     \item dist: Total edit distance.
#'   }
#' @details
#' This function compares read sequences to reference miRNAs and identifies
#' isoforms based on the specified thresholds for indels and mismatches. It calculates the edit distance between the read sequences and the reference miRNAs, and returns a DataFrame containing the detected isoforms that meet the criteria.
#' @examples
#' \dontrun{
#' }
#' @export
find_isoforms <- function(mirnas, reads, max_indel_5p, max_indel_3p, max_snp_5p, max_snp_3p, max_snp_seed, max_snp) {
    .Call(`_risoma_find_isoforms`, mirnas, reads, max_indel_5p, max_indel_3p, max_snp_5p, max_snp_3p, max_snp_seed, max_snp)
}

#' @title Load reads in a FASTQ File
#' @description This function reads a FASTQ file and marks duplicate reads
#'   based on their sequences.
#' @param fq_file Path to the FASTQ file.
#' @return A DataFrame containing two columns:
#'   - `read_seq`: The sequence of the read.
#'   - `tpm`: TPM of the read in the file.
#' @examples
#' \dontrun{
#'   result <- load_reads("example.fq")
#' }
#' @export
load_reads <- function(fq_file) {
    .Call(`_risoma_load_reads`, fq_file)
}

#' @title Load reads in a GZIP-compressed FASTQ File
#' @description This function reads a GZIP-compressed FASTQ file and marks
#'   duplicate reads based on their sequences.
#' @param fq_file Path to the GZIP-compressed FASTQ file.
#' @return A DataFrame containing two columns:
#'   - `read_seq`: The sequence of the read.
#'   - `tpm`: TPM of the read in the file.
#' @examples
#' \dontrun{
#'   result <- load_read_gz("example.fq.gz")
#' }
#' @export
load_reads_gz <- function(fq_file) {
    .Call(`_risoma_load_reads_gz`, fq_file)
}

generate_cigar <- function(seq1, seq2) {
    .Call(`_risoma_generate_cigar`, seq1, seq2)
}

